### ISSUE TITLE
robot-name exercise: Should robot names be case sensitive or insensitive?

### ISSUE THREAD

**Issue by verdammelt** [exercism/x-common#260]
The readme implies that robot names begin with two capital letters. A check of a few tracks show that they are checking for two leading capital letters. Over on the Common Lisp track @wobh asked (2016-11-24) if it should be case insensitive instead (https://github.com/exercism/xlisp/issues/87), @kytrinyx suggested it should be.

The exercise does not seem to have a test data json file yet.

I was tidying up the issues over there in xlisp and though I'd bring this issue up for larger discussion.

**Comment by kytrinyx** [comment]
For what it's worth, I still think they should be case insensitive :)

**Comment by verdammelt** [comment]
Just to be contrary I'll say that they should be all capitals since robots TALK.IN.ALL.CAP-I-TAL.LET-TERS.

So, of course, their names would be as well.

**Comment by kytrinyx** [comment]
@verdammelt ... that would mean that we'd need an error condition for lowercase, right?

**Comment by verdammelt** [comment]
@kytrinyx You have caught me in your inescapable logic! Well played.

Yes, I guess there would need to be a test for some sort of error condition.

**Comment by Insti** [comment]
Since this is one of the early problems, I think it's best if only uppercase letters are allowed.

If not you get the can of worms about whether ab123 and AB123 are the same (as raised in exercism/xlisp#87)

Ruby requires 2 uppercase letters and there is a test for this.

**Comment by kytrinyx** [comment]
@Insti so... tests for some sort of error state on invalid names?

**Comment by Insti** [comment]
@kytrinyx the Robot class generates its own names and they are not assignable, so it makes no sense for it to have name related error state.

The tests should test that the names are valid and such a test* exists in the Ruby version.

* It actually only checks once that one randomly generated name matches the specified format, but this is a deficiency of the test rather than the concept.

**Comment by kytrinyx** [comment]
That makes sense to me.

**Comment by kotp** [comment]
@Insti Since the test is using a random seed for that language, the test may not be deficient, but precisely testing a name possibility that would fail.

If it is infact a deficient test, and does allow what is not wanted, to get through, we should fix that.

**Comment by JPBetley** [comment]
I was lead here trying to fix this ambiguity for the scala exercise.

I looked through the exercism repos to get some data on this issue. Here's what I've found:

- 24 languages have the `robot-name` exercise in their track
- 16 of those languages test for 2 uppercase letters to start the name
- 1 of those languages test for 2 lowercase letters to start the name (php)
- 4 of those languages test any 2 word characters using `\w{2}` (scala, swift, fsharp, objective-c)
- 3 of those languages test for 2 case-insensitive letters (cpp, lua, groovy)

From what I could tell, no test expects error states. The tests just don't pass if the regex isn't valid for the robot name.

Based upon the problem description's examples, and the 2/3 majority of exercises having implemented 2 uppercase letters in tests, I think the problem description should explicitly call for 2 uppercase letters. It also leaves less work to adjust the 1/3 of exercises that differ.
